src/main/java/microsoft/exchange/webservices/data/core/EwsUtilities.java

```diff
@@ -767,60 +702,41 @@ public static String serializeEnum(Object value) {
    * @throws java.text.ParseException the parse exception
    */
+  @SuppressWarnings("unchecked")
   public static <T> T parse(Class<T> cls, String value)
       throws InstantiationException, IllegalAccessException,
       ParseException {
-
     if (cls.isEnum()) {
-      Map<String, String> stringToEnumDict;
-      if (schemaToEnumDictionaries.getMember().containsKey(cls)) {
-        stringToEnumDict = schemaToEnumDictionaries.getMember()
-            .get(cls);
-        if (stringToEnumDict.containsKey(value)) {
-          String strEnumName = stringToEnumDict.get(value);
-          for (Object o : cls.getEnumConstants()) {
-            if (o.toString().equals(strEnumName)) {
-              return (T) o;
-            }
-          }
-          return null;
-        } else {
-          for (Object o : cls.getEnumConstants()) {
-            if (o.toString().equals(value)) {
-              return (T) o;
-            }
-          }
-          return null;
+      final Map<Class<?>, Map<String, String>> member =
+        SCHEMA_TO_ENUM_DICTIONARIES.getMember();
+
+      String val = value;
+      final Map<String, String> stringToEnumDict = member.get(cls);
+      if (stringToEnumDict != null) {
+        final String strEnumName = stringToEnumDict.get(value);
+        if (strEnumName != null) {
+          val = strEnumName;
         }
-      } else {
-        for (Object o : cls.getEnumConstants()) {
-          if (o.toString().equals(value)) {
-            return (T) o;
-          }
+      }
+      for (T o : cls.getEnumConstants()) {
+        if (o.toString().equals(val)) {
+          return o;
         }
-        return null;
       }
-    } else if (cls.isInstance(Integer.valueOf(0)))
-    // else if( cls.isInstance(new Integer(0)))
-    {
-      Object o = null;
-      o = Integer.parseInt(value);
+      return null;
+    } else if (cls.isAssignableFrom(Number.class))  {
+      Integer o = Integer.parseInt(value);
       return (T) o;
-    } else if (cls.isInstance(new Date())) {
-      DateFormat df = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'");
-      df.setTimeZone(TimeZone.getTimeZone("UTC"));
+    } else if (cls.isAssignableFrom(Date.class)) {
+      DateFormat df = createDateFormat(XML_SCHEMA_DATE_TIME_FORMAT);
       return (T) df.parse(value);
-    } else if (cls.isInstance(Boolean.valueOf(false)))
-    // else if( cls.isInstance(new Boolean(false)))
-    {
-      Object o = null;
-      o = Boolean.parseBoolean(value);
+    } else if (cls.isAssignableFrom(Boolean.class)) {
+      Object o = Boolean.parseBoolean(value);
       return (T) o;
-    } else if (cls.isInstance(new String())) {
+    } else if (cls.isAssignableFrom(String.class)) {
       return (T) value;
-    } else if (cls.isInstance(Double.valueOf(0.0))) {
-      Object o = null;
-      o = Double.parseDouble(value);
+    } else if (cls.isAssignableFrom(Double.class)) {
+      Object o = Double.parseDouble(value);
```

Comment:
I don't think that we can remove casting (because we can't do direct cast from primitive type to T), but I've just found a bug: `Number.class` and `Double.class` are processed in the wrong order.